### PR TITLE
Make the binaries contain the commit version

### PR DESCRIPTION
**Pull request by zhujian7**
Signed-off-by: zhujian <jiazhu@redhat.com>

**What this PR does / why we need it**:
After this commit is merged, we can use `command -v` to get the commit id. for example:
```
╰─$ hypershift -v        
hypershift version 2e2f6c3ff15b6abb52d48494c7ad823eed3df764
```

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [ ] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -23,6 +23,7 @@ require (
 	github.com/opencontainers/go-digest v1.0.0
 	github.com/opencontainers/image-spec v1.0.2
 	github.com/openshift/api v0.0.0-20220124143425-d74727069f6f
+	github.com/openshift/build-machinery-go v0.0.0-20220121085309-f94edc2d6874
```
Comment (zhujian7) [inline go.mod:26]: Wait for https://github.com/openshift/build-machinery-go/pull/66 to merge, then I will upgrade it to the latest version

File: pkg/version/version.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	majorFromGit string
+	// minor version
+	minorFromGit string
+	// build date in ISO8601 format, output of $(date -u +'%Y-%m-%dT%H:%M:%SZ')
+	buildDate string
+)
+
+// Get returns the overall codebase version. It's for detecting
+// what code a binary was built from.
+func Get() version.Info {
```
Comment (alvaroaleman) [inline pkg/version/version.go:27]: We don't need any of this, the repo is built with go 1.18. Please get this infro from runtime/debug.ReadBuildInfo (The `Settings` should include git information)

Comment (zhujian7) [reply pkg/version/version.go:27]: Do you mean we use the commit id, which could be gotten from runtime/debug.ReadBuildInfo.Settings[vcs.revision], as the version value, right?

Comment (alvaroaleman) [reply pkg/version/version.go:27]: Sure you can, but not sure if needed. The go version is configured here: https://github.com/openshift/hypershift/blob/main/.ci-operator.yaml#L4

**Comment by zhujian7** [comment]
Now it looks like:
```
╰─$ make build
CGO_ENABLED=0 GO111MODULE=on GOFLAGS=-mod=vendor go build -gcflags=all='-N -l' -o bin/hypershift-operator ./hypershift-operator
CGO_ENABLED=0 GO111MODULE=on GOFLAGS=-mod=vendor go build -gcflags=all='-N -l' -o bin/control-plane-operator ./control-plane-operator
CGO_ENABLED=0 GO111MODULE=on GOFLAGS=-mod=vendor go build -gcflags=all='-N -l' -o bin/hypershift .

╰─$ ./bin/hypershift-operator -v
hypershift-operator version b1247978522cdb047d8779038f7f60107fc851f9
╰─$ ./bin/hypershift -v         
hypershift version b1247978522cdb047d8779038f7f60107fc851f9
╰─$ ./bin/control-plane-operator -v 
control-plane-operator version b1247978522cdb047d8779038f7f60107fc851f9
```

**Comment by enxebre** [comment]
there's two commits with the same message "Make the binaries contain the commit version", can we please squash them?

>BTW, the makefile will [use v[0-9]* to match tags](https://github.com/openshift/build-machinery-go/blob/master/make/lib/golang.mk#L53), but [all tags created in the hypershift repo](https://github.com/openshift/hypershift/tags) are all in format x.y.z, so I propose if it is possible please create tags in the format of vx.y.z.

Is this still relevant, can we update the PR desc?

**Comment by enxebre** [comment]
I'm curious what's the use case that drove you to put this PR @zhujian7?

**Comment by zhujian7** [comment]
> I'm curious what's the use case that drove you to put this PR @zhujian7?

Hi @enxebre, thanks for the review. I am a developer from ACM, now for our upstream build we use the latest hypershift binary to install hypershift operator, and I want to know the commit sha of the hypershift binary.

---
After reconsidering, I think it needs to be careful to make compiling hypershift requires golang 1.18 for this feature, for example ACM still uses golang 1.17, so I want to HOLD the PR. (Sorry @alvaroaleman.)

But if you still think the feature is useful and consider accepting the [first version of the implementation](https://github.com/openshift/hypershift/pull/1324/commits/49e7e01e4910107016d41655c49d8a94c19ec119) (without relying on golang 1.18), I'll be very happy to continue working on it. Thanks.

/hold

**Comment by alvaroaleman** [comment]
> for example ACM still uses golang 1.17

We don't have any testing that uses go 1.17, so if go 1.17 works or doesn't is a matter of luck. This needs to be fixed by ACM.

**Comment by zhujian7** [comment]
> We don't have any testing that uses go 1.17, so if go 1.17 works or doesn't is a matter of luck. This needs to be fixed by ACM.

Thanks @alvaroaleman, good to know. Currently we build it with go 1.17 and it works fine, but you are right, we will try to fix it by ACM

**Comment by zhujian7** [comment]
@alvaroaleman @enxebre golang version problem fixed in ACM, commit message and PR description updated, PTAL.
/hold cancel

**Review by alvaroaleman — APPROVED** [review]
thanks!

**Pull request merged**